Add gravity method super-cycling (#381)
enzo-project/enzo-e: doc/source/design/design-timestep.rst

.. include:: ../roles.incl

.. |tijk| replace:: t\ :sub:`i` :sup:`j,k`

.. toctree::

.. _Adaptive Time-Step Design:

*************************

> Adaptive Time-Step Design

Reviewer: This isn't essential, but we may want to acknowledge that subcycling also exists (e.g. in Grackle and radiative transfer). Maybe just a single sentence noting that this is used as a technique on a method-by-method basis.

These local time-steps can be different for different methods, on
different blocks, and at different points in time.

The simplest time-stepping strategy is to use a constant global
time-step for each cycle, computed as the minimum local time-step over
all methods and blocks in the cycle. This is the approach that Enzo-E
/ Cello used initially. While easy to implement, depending on the
simulation characteristics it can be very inefficient.

Two ways to improve time-stepping performance are to 1) allow adaptive
time-stepping in different regions of the mesh, and 2) allow different
time-steps for different methods.  An example of 1) is the original
ENZO code: ENZO uses different time-steps on different mesh refinement
levels, which can significantly improve performance on mesh
hierarchies with many refinement levels. An example of 2) would be
advancing the gravitational potential, which can be costly due to its
global communication requirements, at a larger time-step than other
methods when possible.

Our plan is to optimize time-stepping in terms of both methods (via
"super-cycling") and individual blocks (via "block-adaptive

> time-stepping").  We implement super-cycling first since it requires

Reviewer: Does block-adaptive time-stepping mean that blocks on the same refinement level can have different timesteps?

are in a sense orthogonal to each other, so both approaches can be
used together as well as individually.

=====================
Super-cycling Gravity
=====================

As a driving use-case, we wish to super-cycle gravity solves with
respect to hydrodynamics. The rationale is that gravity solves are
relatively costly, and gravitational potentials typically evolve at
longer time-scales than hydro fields.

To super-cycle gravity, the user supplies a single integer parameter
:code:`Method : gravity : max_supercycle` that defines how frequently
the :code:`"gravity"` method is to be called with respect to other
methods. The default value of :code:`1` effectively turns off
super-cycling, whereas a value of :math:`k` allows other methods to be
called *up to* :math:`k` times for each call to the gravity
solve. Note that the maximum local gravity time-step restriction on
each block will still always be satisfied, even if the
:code:`max_supercycle` parameter is set to an arbitrarily large value.

For cycles where the gravitational potential solve is not performed,
extrapolations of saved fields are used to approximate the current

> fields.  Which fields to be extrapolated is determined by the

Reviewer: ```suggestion
fields.  The fields to be extrapolated are determined by the
```

           potential, magenta squares represent hydro solves, and blue
           triangles represent accelerations computed from potentials.
           White fields are extrapolated from previously computed
           values.  The left figure shows standard gravity with
           gravity solves every cycle; the figures to the right show
           super-cycled gravity, with the upper-right variation
           showing extrapolated gravitational potentials, and the
           lower-right variation showing extrapolated accelerations.

------------------------
non super-cycled gravity
------------------------

The gravitational potential is used to compute gravitational
accelerations, which are applied to all gravitating particles and
fields. We use a leapfrog integration scheme because it is 2nd order
accurate (small short-scale errors) and symplectic (qualitatively
correct long-scale behavior). This involves computing time-centered
accelerations. Since we start with non-time centered values, we must
either extrapolate the input (density) to the gravity solve, or
extrapolate the output (potential) from the gravity solve. In both

> ENZO and Enzo-E, the extrapolation is done before the gravity solve.

Reviewer: Extrapolating the density involves the `Method : pm_deposit : alpha` parameter, right?

If so then our approach isn't robust in the general case. When I implemented self-gravity source terms for VL+CT, there was a whole saga where I was getting bizare looking results for a standard test problem (i.e. the "stable linear Jeans wave" that is inclined with respect to the grid and for which we can perform an analytic convergence study). Ultimately, the solution was to stop extrapolating. I suspect that we would find similar issues with PPM if we ran a similar test

In that scenario, there were no particles in that case. It's plausible that if particles dominate the short-range gravitational potential that extrapolation might be more robust.

Anyway, we probably don't want to go into detail here, but maybe we add something like the following at the end of this subsection?

```rst
.. caution:

   Some care is needed when applying extrapolation methods. For certain problems, they may not produce meaningful results.
```
